File: src/query.py
"""
A small parser + evaluator for the BibleWorks-style Command Line grammar
(subset: AND '.', phrase ''', OR '/', NOT '!', proximity ';N', compound '(...)',
wildcards * and ?, and morphological lemma@code matching).

Design: parse into an AST of Term / Phrase / And / Or / Not / Proximity nodes,
then evaluate against an in-memory corpus of tagged words (list of dicts with
book/chapter/verse/surface/lemma/bw_code, in document order).
"""
import re
from dataclasses import dataclass, field
from typing import List


# ---------- Wildcard matching ----------

def wildcard_to_regex(pattern: str) -> re.Pattern:
    # '*' -> any sequence, '?' -> any single char. Case-insensitive for text;
    # morph-code matching is handled separately (case sensitive).
    esc = re.escape(pattern).replace(r'\*', '.*').replace(r'\?', '.')
    return re.compile(f'^{esc}$', re.IGNORECASE)


def morph_code_matches(query_code: str, actual_code: str) -> bool:
    """
    query_code may contain '*', '?' wildcards and the grouping operators
    '(' ')' '|' for OR-ing sub-patterns, e.g. '((n-)|(ap))*'
    """
    # Expand '(A|B)' groups into a regex alternation directly.
    # Convert to a regex: outside-of-group '*'/'?' become wildcards,
    # parentheses/| pass through as regex groups (already valid regex syntax).
    out = []
    i = 0
    while i < len(query_code):
        c = query_code[i]
        if c == '*':
            out.append('.*')
        elif c == '?':
            out.append('.')
        elif c in '()|':
            out.append(c)
        else:
            out.append(re.escape(c))
        i += 1
    pattern = '^' + ''.join(out) + '$'
    return re.match(pattern, actual_code, re.IGNORECASE) is not None


# ---------- AST ----------

@dataclass
class Word:
    """A single search term: plain text, or lemma@morphcode for morph versions."""
    raw: str

    def lemma_and_code(self):
        if '@' in self.raw:
            lemma, code = self.raw.split('@', 1)
            return lemma, code
        return self.raw, None


@dataclass
class Phrase:
    words: List[Word]
    max_gap: int = 0  # '*N' between words allows up to N intervening words; 0 = exact phrase


@dataclass
class And:
    positive: List  # terms required present
    negative: List  # terms required absent ('!')
    proximity: int = None  # ';N' verse window; None = same verse


@dataclass
class Or:
    options: List


@dataclass
class Not:
    inner: object


# ---------- Tokenizing / parsing the command line ----------
# Supported subset for this prototype:
#   .term1 term2!term3;N      (AND, with NOT and proximity)
#   /term1 term2              (OR)
#   'term1 term2 *N term3     (PHRASE, with gap)
#   (subexpr).(subexpr)       (compound AND of two parenthesized groups)
#   (subexpr)/(subexpr)       (compound OR)
#   (subexpr).!(subexpr)      (compound AND-NOT)

def parse(cmdline: str):
    cmdline = cmdline.strip()
    if cmdline.startswith('('):
        return _parse_compound(cmdline)
    if cmdline.startswith('.'):
        return _parse_and(cmdline[1:])
    if cmdline.startswith('/'):
        return _parse_or(cmdline[1:])
    if cmdline.startswith("'"):
        return _parse_phrase(cmdline[1:])
    raise ValueError(f"Unrecognized command line start: {cmdline!r}")


def _split_top_level(s: str, sep_chars: str):
    """Split on sep_chars, but not inside parentheses."""
    parts, depth, cur = [], 0, ''
    for c in s:
        if c == '(':
            depth += 1
        elif c == ')':
            depth -= 1
        if c in sep_chars and depth == 0:
            parts.append(cur); cur = ''
        else:
            cur += c
    parts.append(cur)
    return parts


def _parse_and(s: str) -> And:
    prox = None
    if ';' in s:
        s, prox_str = s.rsplit(';', 1)
        prox = int(prox_str)
    tokens = s.split()
    pos, neg = [], []
    for t in tokens:
        if t.startswith('!'):
            neg.append(Word(t[1:]))
        else:
            pos.append(Word(t))
    return And(pos, neg, prox)


def _parse_or(s: str) -> Or:
    return Or([Word(t) for t in s.split()])


def _parse_phrase(s: str) -> Phrase:
    tokens = s.split()
    words, gap = [], 0
    i = 0
    while i < len(tokens):
        t = tokens[i]
        m = re.match(r'^\*(\d*)$', t)
        if m:
            gap = int(m.group(1)) if m.group(1) else 99999  # bare '*' = any number of words
        else:
            words.append(Word(t))
        i += 1
    return Phrase(words, gap)


def _parse_compound(s: str):
    # find matching close-paren for the first group
    depth = 0
    for i, c in enumerate(s):
        if c == '(':
            depth += 1
        elif c == ')':
            depth -= 1
            if depth == 0:
                first = s[1:i]
                rest = s[i+1:]
                break
    else:
        raise ValueError("Unbalanced parentheses")
    left = parse_subexpr(first)
    if not rest:
        return left
    if rest.startswith('.!'):
        frag = rest[2:]
        if frag.startswith('(') and frag.endswith(')'):
            frag = frag[1:-1]
        right = parse_subexpr(frag)
        return And([left], [right])
    if rest.startswith('.'):
        right_str = rest[1:]
        right = parse_subexpr(right_str)
        return And([left, right], [])
    if rest.startswith('/'):
        right = parse_subexpr(rest[1:])
        return Or([left, right])
    if rest[0].isdigit() or rest.startswith('.') is False:
        # proximity form like (...)N(...)
        m = re.match(r'^(\d+)', rest)
        if m:
            n = int(m.group(1))
            right = parse_subexpr(rest[m.end():])
            return And([left, right], [], proximity=n)
    raise ValueError(f"Unsupported compound continuation: {rest!r}")


def parse_subexpr(s: str):
    """Parse a fragment that may itself start with ( . / ' or be bare (treated as AND term list)."""
    s = s.strip()
    if s.startswith('('):
        return _parse_compound(s)
    if s.startswith('.'):
        return _parse_and(s[1:])
    if s.startswith('/'):
        return _parse_or(s[1:])
    if s.startswith("'"):
        return _parse_phrase(s[1:])
    # bare term list inside parens (e.g. 'barnabas' in (barnabaj))
    return _parse_and('.' + s if not s.startswith('.') else s)

File: docs/js/query.js
// query.js — faithful JS port of src/query.py
// A small parser + evaluator for the BibleWorks-style Command Line grammar
// (subset: AND '.', phrase ''', OR '/', NOT '!', proximity ';N', compound '(...)',
// wildcards * and ?, and morphological lemma@code matching).

(function (global) {
  'use strict';

  function reEscape(s) {
    return s.replace(/[.*+?^${}()|[\]\\]/g, '\\$&');
  }

  // '*' -> any sequence, '?' -> any single char. Case-insensitive.
  function wildcardToRegex(pattern) {
    const esc = reEscape(pattern).replace(/\\\*/g, '.*').replace(/\\\?/g, '.');
    return new RegExp('^' + esc + '$', 'i');
  }

  // query_code may contain '*', '?' wildcards and the grouping operators
  // '(' ')' '|' for OR-ing sub-patterns, e.g. '((n-)|(ap))*'
  function morphCodeMatches(queryCode, actualCode) {
    let out = [];
    for (let i = 0; i < queryCode.length; i++) {
      const c = queryCode[i];
      if (c === '*') out.push('.*');
      else if (c === '?') out.push('.');
      else if (c === '(' || c === ')' || c === '|') out.push(c);
      else out.push(reEscape(c));
    }
    const pattern = '^' + out.join('') + '$';
    try {
      return new RegExp(pattern, 'i').test(actualCode == null ? '' : actualCode);
    } catch (e) {
      return false;
    }
  }

  class Word {
    constructor(raw) { this.raw = raw; }
    lemmaAndCode() {
      if (this.raw.includes('@')) {
        const idx = this.raw.indexOf('@');
        return [this.raw.slice(0, idx), this.raw.slice(idx + 1)];
      }
      return [this.raw, null];
    }
  }

  class Phrase {
    constructor(words, maxGap) { this.words = words; this.max_gap = maxGap || 0; }
  }

  class And {
    constructor(positive, negative, proximity) {
      this.positive = positive; this.negative = negative;
      this.proximity = (proximity === undefined) ? null : proximity;
    }
  }

  class Or {
    constructor(options) { this.options = options; }
  }

  // Python str.split() semantics: split on runs of whitespace, drop empties,
  // and return [] for an empty/whitespace-only string.
  function pysplit(s) {
    const t = s.trim();
    if (!t) return [];
    return t.split(/\s+/);
  }

  function parse(cmdline) {
    cmdline = cmdline.trim();
    if (cmdline.startsWith('(')) return parseCompound(cmdline);
    if (cmdline.startsWith('.')) return parseAnd(cmdline.slice(1));
    if (cmdline.startsWith('/')) return parseOr(cmdline.slice(1));
    if (cmdline.startsWith("'")) return parsePhrase(cmdline.slice(1));
    throw new Error(`Unrecognized command line start: ${JSON.stringify(cmdline)}`);
  }

  function splitTopLevel(s, sepChars) {
    const parts = [];
    let depth = 0, cur = '';
    for (const c of s) {
      if (c === '(') depth += 1;
      else if (c === ')') depth -= 1;
      if (sepChars.includes(c) && depth === 0) {
        parts.push(cur); cur = '';
      } else {
        cur += c;
      }
    }
    parts.push(cur);
    return parts;
  }

  function parseAnd(s) {
    let prox = null;
    const semiIdx = s.lastIndexOf(';');
    if (semiIdx !== -1) {
      const proxStr = s.slice(semiIdx + 1);
      s = s.slice(0, semiIdx);
      prox = parseInt(proxStr, 10);
    }
    const tokens = pysplit(s);
    const pos = [], neg = [];
    for (const t of tokens) {
      if (t.startsWith('!')) neg.push(new Word(t.slice(1)));
      else pos.push(new Word(t));
    }
    return new And(pos, neg, prox);
  }

  function parseOr(s) {
    return new Or(pysplit(s).map((t) => new Word(t)));
  }

  function parsePhrase(s) {
    const tokens = pysplit(s);
    const words = [];
    let gap = 0;
    for (const t of tokens) {
      const m = /^\*(\d*)$/.exec(t);
      if (m) {
        gap = m[1] ? parseInt(m[1], 10) : 99999; // bare '*' = any number of words
      } else {
        words.push(new Word(t));
      }
    }
    return new Phrase(words, gap);
  }

  function parseCompound(s) {
    let depth = 0, first = null, rest = null;
    for (let i = 0; i < s.length; i++) {
      const c = s[i];
      if (c === '(') depth += 1;
      else if (c === ')') {
        depth -= 1;
        if (depth === 0) {
          first = s.slice(1, i);
          rest = s.slice(i + 1);
          break;
        }
      }
    }
    if (first === null) throw new Error('Unbalanced parentheses');
    const left = parseSubexpr(first);
    if (!rest) return left;
    if (rest.startsWith('.!')) {
      let frag = rest.slice(2);
      if (frag.startsWith('(') && frag.endsWith(')')) frag = frag.slice(1, -1);
      const right = parseSubexpr(frag);
      return new And([left], [right]);
    }
    if (rest.startsWith('.')) {
      const right = parseSubexpr(rest.slice(1));
      return new And([left, right], []);
    }
    if (rest.startsWith('/')) {
      const right = parseSubexpr(rest.slice(1));
      return new Or([left, right]);
    }
    if (/^[0-9]/.test(rest[0] || '') || !rest.startsWith('.')) {
      const m = /^(\d+)/.exec(rest);
      if (m) {
        const n = parseInt(m[1], 10);
        const right = parseSubexpr(rest.slice(m[0].length));
        return new And([left, right], [], n);
      }
    }
    throw new Error(`Unsupported compound continuation: ${JSON.stringify(rest)}`);
  }

  function parseSubexpr(s) {
    s = s.trim();
    if (s.startsWith('(')) return parseCompound(s);
    if (s.startsWith('.')) return parseAnd(s.slice(1));
    if (s.startsWith('/')) return parseOr(s.slice(1));
    if (s.startsWith("'")) return parsePhrase(s.slice(1));
    return parseAnd(s.startsWith('.') ? s : '.' + s);
  }

  global.QueryLib = {
    Word, Phrase, And, Or,
    wildcardToRegex, morphCodeMatches,
    parse, parseSubexpr, splitTopLevel,
  };
})(window);
